### ISSUE TITLE
Accessing files in IGV (Range header support)

### ISSUE THREAD

**Issue by holtgrewe** [UtrechtUniversity/davrods#9]
Hi, I have problems with accessing BAM files through HTTPS+Davrods in IGV.

Does Davrods support HTTP range queries?

I'm getting the following access/error logs

```
==> /var/log/apache2/error.log <==
[Thu Aug 24 14:34:28.938949 2017] [dav:error] [pid 2451:tid 139841358481152] [client 172.16.128.8:57958] Unable to deliver content.  [500, #0]
[Thu Aug 24 14:34:28.939049 2017] [dav:error] [pid 2451:tid 139841358481152] (104)Connection reset by peer: [client 172.16.128.8:57958] Could not write contents to filter.  [500, #0]

==> /var/log/apache2/other_vhosts_access.log <==
cubi-webdav-omics-prototype.bihealth.org:443 172.16.128.8 - holtgrem [24/Aug/2017:14:32:27 +0200] "GET /omicsZone/projects/project8/bio_samples/33900/33900-N1/33900-N1-DNA1/33900-N1-DNA1-WES1/2017-08-23/GRCh37/ngs_mapping/out/bwa.33900-N1-DNA1-WES1.bam.bai HTTP/1.1" 500 519832 "-" "IGV Version 2.3.97 (157)07/10/2017 10:52 PM"
cubi-webdav-omics-prototype.bihealth.org:443 172.16.128.8 - holtgrem [24/Aug/2017:14:34:35 +0200] "HEAD /omicsZone/projects/project8/bio_samples/33900/33900-N1/33900-N1-DNA1/33900-N1-DNA1-WES1/2017-08-23/GRCh37/ngs_mapping/out/bwa.33900-N1-DNA1-WES1.bam.bai HTTP/1.1" 200 2352 "-" "IGV Version 2.3.97 (157)07/10/2017 10:52 PM"
cubi-webdav-omics-prototype.bihealth.org:443 172.16.128.8 - holtgrem [24/Aug/2017:14:34:35 +0200] "HEAD /omicsZone/projects/project8/bio_samples/33900/33900-N1/33900-N1-DNA1/33900-N1-DNA1-WES1/2017-08-23/GRCh37/ngs_mapping/out/bwa.33900-N1-DNA1-WES1.bam HTTP/1.1" 200 277 "-" "IGV Version 2.3.97 (157)07/10/2017 10:52 PM"
```

IGV is giving me

```
Error loading https://cubi-webdav-omics-prototype.bihealth.org/omicsZone/projects/project8/bio_samples/33900/33900-N1/33900-N1-DNA1/33900-N1-DNA1-WES1/2017-08-23/GRCh37/ngs_mapping/out/bwa.33900-N1-DNA1-WES1.bam: Premature EOF. Expected 4 but only received 0; BinaryCodec in readmode; streamed file (filename not available)
```

**Comment by cjsmeele** [comment]
> Does Davrods support HTTP range queries?

Hi, thanks for reporting this! Partial uploads are supported, but downloads with ranges not yet.
Currently the entire file is delivered regardless of any Range headers, which is rightly confusing for clients.

**Comment by holtgrewe** [comment]
It would be absolutely wonderful if I could point IGV at my iRODS file through davrods!

**Comment by holtgrewe** [comment]
Hi, is there anything I can do to help out with this?

**Comment by cjsmeele** [comment]
If you would like to work on this issue, you could create a fork and a PR when you're done. Contributions are appreciated!

I'm working on some other priorities right now, so I'll pick this up later otherwise.

**Comment by holtgrewe** [comment]
Could you point me at the part of the API to provide for supporting random acccess in and push this throught to the client?

This appears non-trivial to me at the moment...

**Comment by cjsmeele** [comment]
> Could you point me at the part of the API to provide for supporting random acccess in and push this throught to the client?

The affected functions in Davrods are `dav_repo_set_headers` and `deliver_file` in `repo.c`.
Based on the Range header in the request struct `r` we would seek/read/send the requested parts of an iRODS object. When multiple ranges are requested at once we would need to output in multipart format (set appropriate headers and insert boundaries according to [RFC7233](https://tools.ietf.org/html/rfc7233#appendix-A)).
For parsing the Range header and generating the multipart format we'd want to reuse existing Apache functions as much as possible.

>This appears non-trivial to me at the moment

If it's too much that is totally fine - I can work on it at a later time.
If you do decide to work on it, please let me know!

**Comment by holtgrewe** [comment]
Hi, I took a stab at this but the work required here appears to be more involved than I thought initially.

Apparently, the core of apache uses something called [byte range filters](https://httpd.apache.org/docs/2.4/developer/filters.html) which implement the range request feature. I have no idea  how this would interplay with the way Davrods currently handles data sending with the bucket brigade feature of Apache...

Would you consider an implementation that only supports single range requests for inclusion? I think this is something that I could make work in a relatively short time.

FWIW, I found this documentation for range requests useful:

- https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests

**Comment by cjsmeele** [comment]
We've looked into this, and came to this conclusion:

1. Davrods currently does not advertise `Range` header support -- no `Accept-Ranges` header is sent. Clients should handle this by either not sending range requests or by accepting the complete resource.
2. Current server behavior seems to be according to spec: Ranges are not supported so the whole file is sent back.
3. We very much want to support Range requests
4. If we were to advertise `Accept-Ranges` to clients when we do not have complete support then we would break HTTP specifications (multiple ranges and conditional ranges must also be supported).

Considering also the wishes of the product owner (@tsmeele) to remain fully compliant, we will not release partial support for ranges -- only a version that behaves according to spec like Apache's implementation.

We hope you understand this point of view. If this feature is urgent for your use case you could write a partial implemention in a fork, until we have added complete range support.

**Comment by holtgrewe** [comment]
Hi Chris. Thank you for looking into this.

Do you guys see a way to piggy back on the existing Apache byte range filter? Otherwise, I think we will need to reimplement this...

**Comment by cjsmeele** [comment]
We cannot use the filter directly, but for parsing and filtering the range header I think we can call existing functions (`ap_set_byterange`) in the byterange filter. This also seems to take care of conditional ranges.
We may need to split this off into Davrods (similar to what we did for mod\_dav\_lock) because the functions don't seem to be exposed to other modules.
If this existing code proves to be usable, I think the Davrods-specific code to develop will be:
- Setting the right headers
- Handling the parsed range information and seek/read/write the requested parts.
- Output multipart boundaries if necessary

**Comment by cjsmeele** [comment]
This feature is included in [4.2.2_1.4.0](https://github.com/UtrechtUniversity/davrods/releases/tag/4.2.2_1.4.0)
